svgnative/ports/cairo/CairoSVGRenderer.cpp

```diff
@@ -0,0 +1,473 @@
+            cairo_set_fill_rule (mCairo, CAIRO_FILL_RULE_EVEN_ODD);
+            break;
+        case WindingRule::kNonZero:
+        default:
+            cairo_set_fill_rule (mCairo, CAIRO_FILL_RULE_WINDING);
+        }
+
+        cairo_new_path (mCairo);
+        appendCairoSvgPath( mCairo, path );
+        cairo_fill (mCairo);
+    }
+    if (strokeStyle.hasStroke)
+    {
+        const auto& color = boost::get<Color>(strokeStyle.paint);
+        cairo_set_source_rgba(mCairo,
+                              color[0],
+                              color[1],
+                              color[2],
+                              color[3] * strokeStyle.strokeOpacity * alpha );
+
+        cairo_set_line_width(mCairo, strokeStyle.lineWidth);
+
+        switch (strokeStyle.lineCap) {
+        case LineCap::kRound:
+            cairo_set_line_cap(mCairo, CAIRO_LINE_CAP_ROUND);
+            break;
+        case LineCap::kSquare:
+            cairo_set_line_cap(mCairo, CAIRO_LINE_CAP_SQUARE);
+            break;
+        case LineCap::kButt:
+        default:
+            cairo_set_line_cap(mCairo, CAIRO_LINE_CAP_BUTT);
+        }
+
+        switch (strokeStyle.lineJoin) {
+	case LineJoin::kRound:
+	    cairo_set_line_join(mCairo, CAIRO_LINE_JOIN_ROUND);
+            break;
+	case LineJoin::kBevel:
+	    cairo_set_line_join(mCairo, CAIRO_LINE_JOIN_BEVEL);
+            break;
+	case LineJoin::kMiter:
+	default:
+	    cairo_set_line_join(mCairo, CAIRO_LINE_JOIN_MITER);
+        }
+
+        // dashArray is not supported in Skia yet (2019/04/28)
+        // copied from String backend
+        double *dashes = NULL;
+        if (!strokeStyle.dashArray.empty()) {
+            dashes = (double*)malloc( strokeStyle.dashArray.size() * sizeof(double) );
+            for (size_t i = 0; i < strokeStyle.dashArray.size(); i ++ )
+                dashes[i] = strokeStyle.dashArray[i];
+            cairo_set_dash(mCairo, dashes, strokeStyle.dashArray.size(), strokeStyle.dashOffset );
+        }
+
+        cairo_new_path (mCairo);
+        appendCairoSvgPath( mCairo, path );
+        cairo_stroke (mCairo);
+        if (dashes)
+            free (dashes);
+    }
+    Restore();
+}
+
+void CairoSVGRenderer::DrawImage(
+    const ImageData& image, const GraphicStyle& graphicStyle, const Rect& clipArea, const Rect& fillArea)
+{
+    double alpha = getAlphaProduct( alphas );
+
+    SVG_ASSERT(mCairo);
+    Save(graphicStyle);
+    cairo_new_path (mCairo);
+    cairo_rectangle (mCairo, clipArea.x, clipArea.y, clipArea.width, clipArea.height);
+    cairo_clip(mCairo);
+
+    float w = image.Width();
+    float h = image.Height();
+    cairo_surface_t *mImageData = (static_cast<const CairoSVGImageData&>(image)).mImageData;
+
+    cairo_translate (mCairo, fillArea.x, fillArea.y);
+    cairo_scale (mCairo, fillArea.width / w, fillArea.height / h );
+    cairo_set_source_surface (mCairo, mImageData, 0, 0);
+    cairo_paint_with_alpha (mCairo, alpha);
```

Comment:
This is accumulative as far as I remember. No need to multiply alpha IIRC.

Reply:
OK, I would use opacity directly.